### ISSUE TITLE
Get commit messages in the codecommit/repository parser

### ISSUE THREAD

**Issue by elChapoSing** [arabold/aws-to-slack#51]
Everything works as expected but it would be even greater to be able to add the commit message itself to the slack card.
Not so much an issue but rather a feature request.

**Comment by elChapoSing** [comment]
Doesn't look like the commit message is in the cloudwatch event message so bit of a no-go I guess.
Could it be loaded from the `detail.commitid` and `detail.repositoryid` ?

**Comment by thesuavehog** [comment]
Amazon had some sample code (which I've modified quite a bit) that demonstrates how to pull this info. It's not trivial as you have to go looking for it in CodeCommit vs just pulling it off the notification.
```
# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
# SPDX-License-Identifier: MIT-0

import boto3
import os
import json
import pprint
import datetime

# Define the environment variables for repository branch name and region
REGION = os.getenv('AWS_REGION')
#MAIN_BRANCH_NAME = os.getenv('MAIN_BRANCH_NAME')
#REPOSITORY_NAME = os.getenv('REPOSITORY_NAME')

codecommit = boto3.client('codecommit')

def publish(repository, message):
  SNSTopicArn = os.getenv('SNS_TOPIC_ARN')
  SNSClient = boto3.client('sns', region_name=REGION)
  SNSClient.publish(
      TopicArn=SNSTopicArn,
      Subject = 'CodeCommit Update - Repository: {0}'.format(repository),
      Message = message
  )

def getFileDifferences(repository_name, lastCommitID, previousCommitID):
    response = None

    if previousCommitID != None:
        response = codecommit.get_differences(
            repositoryName=repository_name,
            beforeCommitSpecifier=previousCommitID,
            afterCommitSpecifier=lastCommitID
        )
    else:
        # The case of getting initial commit (Without beforeCommitSpecifier)
        response = codecommit.get_differences(
            repositoryName=repository_name,
            afterCommitSpecifier=lastCommitID
        )

    differences = []

    if response == None:
        return differences

    while "nextToken" in response:
        response = codecommit.get_differences(
            repositoryName=repository_name,
            beforeCommitSpecifier=previousCommitID,
            afterCommitSpecifier=lastCommitID,
            nextToken=response["nextToken"]
        )
        differences += response.get("differences", [])
    else:
        differences += response["differences"]

    return differences

def getDiffChangeTypeMessage(changeType):
    type = {
        'M': 'Modification',
        'D': 'Deletion',
        'A': 'Addition'
    }
    return type[changeType]

def getLastCommitID(repository, branch="master"):
    response = codecommit.get_branch(
        repositoryName=repository,
        branchName=branch
    )
    commitId = response['branch']['commitId']
    return commitId

def getLastCommitLog(repository, commitId):
    response = codecommit.get_commit(
        repositoryName=repository,
        commitId=commitId
    )
    return response['commit']


def getMessageText(differences, lastCommit):
    text = ''
    text += 'Commit ID: {0}\n'.format(lastCommit['commitId'])
    text += 'Author: {0} ({1}) - {2} UTC\n'.format(lastCommit['author']['name'], lastCommit['author']['email'], datetime.datetime.fromtimestamp(float(lastCommit['author']['date'].split(' ')[0])).strftime("%Y-%m-%d %I:%M:%S %p"))
    text += 'Message:\n{0}\n\n'.format(lastCommit['message'])
    for diff in differences:
        if 'afterBlob' in diff:
            text += '{0} {1}\n'.format(diff['changeType'], diff['afterBlob']['path'])
            #text += '{0} {1} - Blob ID: {2}\n'.format(diff['changeType'], diff['afterBlob']['path'], diff['afterBlob']['blobId'])
        #if 'beforeBlob' in diff:
        #    text += '{0} {1} - Blob ID: {2}\n'.format(diff['changeType'], diff['beforeBlob']['path'], diff['beforeBlob']['blobId'])

    return text

def lambda_handler(event, context):
    #print json.dumps(event, indent=3)
    repository = event['detail']['repositoryName']
    branch = "(unknown)"
    if event['detail']['referenceType'] == "branch":
        branch = event['detail']['referenceName']
    else:
        return

    print('Notify for change on {0}:{1} by {2}'.format(repository, branch, event['detail']['callerUserArn'].split(':')[5].split('/')[1]))

    try:
        lastCommitID = getLastCommitID(repository, branch)
        lastCommit = getLastCommitLog(repository, lastCommitID)

        previousCommitID = None
        if len(lastCommit['parents']) > 0:
            previousCommitID = lastCommit['parents'][0]

        print('lastCommitID: {0} previousCommitID: {1}'.format(lastCommitID, previousCommitID))

        differences = getFileDifferences(repository, lastCommitID, previousCommitID)
        messageText =  'Repository: {0}\n'.format(repository)
        messageText += 'Branch: {0}\n'.format(branch)
        messageText += getMessageText(differences, lastCommit)

        return publish(repository, messageText)

    except Exception as e:
        print(e)
        print('Error getting repository {}. Make sure it exists and that your repository is in the same region as this function.'.format(repository))
        raise e
```

**Comment by thesuavehog** [comment]
Sorry about the formatting.... not sure how to get code to display properly in this editor,

**Comment by lorengordon** [comment]
use triple backticks to format multi-line code blocks

```
    ```
    code
    ```
```

**Comment by thesuavehog** [comment]
Thanks!

Note: the code I copied is used to push a notification to SNS but you can edit as required.

**Comment by lorengordon** [comment]
Yeah, it's also python, where this project uses javascript. Wouldn't really be all that difficult to add, I don't think. I can see the value for sure. While I contributed the codecommit parser, I just don't have time to investigate this at the moment. So I'd say dive in and submit a PR! :D

**Comment by thesuavehog** [comment]
Yeah - I know it's different language but figured it might help someone as a reference if they're looking to move on it since it will take me a while to find the time to code it up. =)

**Comment by elChapoSing** [comment]
thks for pointing me in the right direction. might give it a try if I find the time.

**Comment by elChapoSing** [comment]
@lorengordon if you have 2 seconds to check out my pull request ? I'm getting an error that I don't understand and you must have a lot more experience than me in using nodejs with lambda.

**Comment by lorengordon** [comment]
@elChapoSing Unfortunately my experience is mostly superficial. I can copy/paste with the best of them, but Node/JavaScript is a near total mystery to me. So, mostly just guessing here, but this project has moved to the new async methods available in nodejs 8.10, so maybe the promise approach in your patch is not compatible. If it works locally, I'd suspect a delta in the pkgs in your working environment vs the pkgs listed as dependencies in this project...

**Comment by elChapoSing** [comment]
solved by PR#53

**Comment by saqib-alharab** [comment]
> Amazon had some sample code (which I've modified quite a bit) that demonstrates how to pull this info. It's not trivial as you have to go looking for it in CodeCommit vs just pulling it off the notification.
> 
> ```
> # Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
> # SPDX-License-Identifier: MIT-0
> 
> import boto3
> import os
> import json
> import pprint
> import datetime
> 
> # Define the environment variables for repository branch name and region
> REGION = os.getenv('AWS_REGION')
> #MAIN_BRANCH_NAME = os.getenv('MAIN_BRANCH_NAME')
> #REPOSITORY_NAME = os.getenv('REPOSITORY_NAME')
> 
> codecommit = boto3.client('codecommit')
> 
> def publish(repository, message):
>   SNSTopicArn = os.getenv('SNS_TOPIC_ARN')
>   SNSClient = boto3.client('sns', region_name=REGION)
>   SNSClient.publish(
>       TopicArn=SNSTopicArn,
>       Subject = 'CodeCommit Update - Repository: {0}'.format(repository),
>       Message = message
>   )
> 
> def getFileDifferences(repository_name, lastCommitID, previousCommitID):
>     response = None
> 
>     if previousCommitID != None:
>         response = codecommit.get_differences(
>             repositoryName=repository_name,
>             beforeCommitSpecifier=previousCommitID,
>             afterCommitSpecifier=lastCommitID
>         )
>     else:
>         # The case of getting initial commit (Without beforeCommitSpecifier)
>         response = codecommit.get_differences(
>             repositoryName=repository_name,
>             afterCommitSpecifier=lastCommitID
>         )
> 
>     differences = []
> 
>     if response == None:
>         return differences
> 
>     while "nextToken" in response:
>         response = codecommit.get_differences(
>             repositoryName=repository_name,
>             beforeCommitSpecifier=previousCommitID,
>             afterCommitSpecifier=lastCommitID,
>             nextToken=response["nextToken"]
>         )
>         differences += response.get("differences", [])
>     else:
>         differences += response["differences"]
> 
>     return differences
> 
> def getDiffChangeTypeMessage(changeType):
>     type = {
>         'M': 'Modification',
>         'D': 'Deletion',
>         'A': 'Addition'
>     }
>     return type[changeType]
> 
> def getLastCommitID(repository, branch="master"):
>     response = codecommit.get_branch(
>         repositoryName=repository,
>         branchName=branch
>     )
>     commitId = response['branch']['commitId']
>     return commitId
> 
> def getLastCommitLog(repository, commitId):
>     response = codecommit.get_commit(
>         repositoryName=repository,
>         commitId=commitId
>     )
>     return response['commit']
> 
> 
> def getMessageText(differences, lastCommit):
>     text = ''
>     text += 'Commit ID: {0}\n'.format(lastCommit['commitId'])
>     text += 'Author: {0} ({1}) - {2} UTC\n'.format(lastCommit['author']['name'], lastCommit['author']['email'], datetime.datetime.fromtimestamp(float(lastCommit['author']['date'].split(' ')[0])).strftime("%Y-%m-%d %I:%M:%S %p"))
>     text += 'Message:\n{0}\n\n'.format(lastCommit['message'])
>     for diff in differences:
>         if 'afterBlob' in diff:
>             text += '{0} {1}\n'.format(diff['changeType'], diff['afterBlob']['path'])
>             #text += '{0} {1} - Blob ID: {2}\n'.format(diff['changeType'], diff['afterBlob']['path'], diff['afterBlob']['blobId'])
>         #if 'beforeBlob' in diff:
>         #    text += '{0} {1} - Blob ID: {2}\n'.format(diff['changeType'], diff['beforeBlob']['path'], diff['beforeBlob']['blobId'])
> 
>     return text
> 
> def lambda_handler(event, context):
>     #print json.dumps(event, indent=3)
>     repository = event['detail']['repositoryName']
>     branch = "(unknown)"
>     if event['detail']['referenceType'] == "branch":
>         branch = event['detail']['referenceName']
>     else:
>         return
> 
>     print('Notify for change on {0}:{1} by {2}'.format(repository, branch, event['detail']['callerUserArn'].split(':')[5].split('/')[1]))
> 
>     try:
>         lastCommitID = getLastCommitID(repository, branch)
>         lastCommit = getLastCommitLog(repository, lastCommitID)
> 
>         previousCommitID = None
>         if len(lastCommit['parents']) > 0:
>             previousCommitID = lastCommit['parents'][0]
> 
>         print('lastCommitID: {0} previousCommitID: {1}'.format(lastCommitID, previousCommitID))
> 
>         differences = getFileDifferences(repository, lastCommitID, previousCommitID)
>         messageText =  'Repository: {0}\n'.format(repository)
>         messageText += 'Branch: {0}\n'.format(branch)
>         messageText += getMessageText(differences, lastCommit)
> 
>         return publish(repository, messageText)
> 
>     except Exception as e:
>         print(e)
>         print('Error getting repository {}. Make sure it exists and that your repository is in the same region as this function.'.format(repository))
>         raise e
> ```

Hi 

I need to get the file to which the last commit was done

How to do that?
